### ISSUE TITLE
Explore python's pex for deployment

### ISSUE THREAD

**Issue by tanin47** [google/clusterfuzz-tools#157]
Using `pip` for deployment has been proven problematic because Goobuntu is shipped with pip 1.5.4; it is known to be buggy when installing all dependencies needed by clusterfuzz-tools.

Using [pex](https://github.com/pantsbuild/pex) (from Twitter :P) enables us to ship python as a single binary. And we can ship the binary in a pip package. Pip 1.5.4 should have no trouble install a pip package without any dependency.

**Comment by tanin47** [comment]
End users only need python to execute a pex-packaged binary.

**Comment by oliverchang** [comment]
Chiming in here. There's also https://github.com/google/subpar :)

**Comment by tanin47** [comment]
Hmmm... `subpar` seems to depend on Bazel. It seems `subpar` can be used independently (from reading its code); but the instruction isn't written in the README. That's weird.

But moving to Bazel is not a bad idea though. I hate the fact that we depend on the relativeness of paths for running tests. We should actually use Bazel in our clusterfuzz repo as well.

---------

`pex -r prod_requirements.txt -e clusterfuzz.main:execute -o clusterfuzzpex .`  generates `clusterfuzzpex` that works perfectly ok.

**Comment by tanin47** [comment]
`bazel` seems to work. But I don't think we can use `subpar` because `blaze` and `subpar` don't support installing packages from pip. The main problem is that a pypi (or pip) package has its own pypi dependencies. We need to pull all of those recursive dependencies. Therefore, it's important to use pip.

Here's [an example how to include a single pip package](http://tanin.nanakorn.com/blogs/370). 

From googling around, we want to support it (https://github.com/bazelbuild/bazel/issues/699), but none has implemented it yet. I tried https://github.com/gengo/rules_pypi but can't get it to work.

`pex` supports pip natively... so we will go with `pex`.

**Comment by tanin47** [comment]
After thinking about it for a while, specifying the pypi dependencies should be in the build tool (not the packaging tool); That's exactly what `pants` + `pex` does e.g http://www.pantsbuild.org/3rdparty_py.html.

But I think we probably shouldn't use `pants`. Therefore, we will hard-code these dependencies with `new_http_archive`:

```
appdirs==1.4.3
asn1crypto==0.22.0
backports.shutil-get-terminal-size==1.0.0
boto==2.42.0
cffi==1.10.0
clusterfuzz==0.2.2rc4
crcmod==1.7
cryptography==1.8.1
enum34==1.1.6
funcsigs==1.0.2
gcs-oauth2-boto-plugin==1.14
google-apitools==0.5.3
gsutil==4.23
httplib2==0.10.3
idna==2.5
ipaddress==1.0.18
mock==2.0.0
oauth2client==2.2.0
packaging==16.8
pbr==2.1.0
psutil==5.2.0
pyasn1==0.2.3
pyasn1-modules==0.0.8
pycparser==2.17
pyOpenSSL==16.2.0
pyparsing==2.2.0
python-gflags==3.1.1
PyYAML==3.12
requests==2.13.0
retry-decorator==1.1.0
rsa==3.4.2
six==1.10.0
SocksiPy-branch==1.1
urlfetch==1.0.2
xvfbwrapper==0.2.9
```

**Comment by tanin47** [comment]
It still doesn't work because some pip packages need to be compiled.

**Comment by tanin47** [comment]
`pants` + `pex` works fine (including testing and building binary). But `pants` doesn't work on windows. This is probably a deal breaker.

On another hand, `bazel` will work on windows. But it doesn't support packages from pip. And, if we hack something together, it's risky that `subpar` won't package stuffs correctly. Writing an extension on this is not trivial, and we need to know bazel's internals.

**Comment by tanin47** [comment]
Nevermind, thanks to this comment: https://github.com/bazelbuild/bazel/issues/699#issuecomment-262091510

`pants` doesn't support windows. That's a big no.

I have successfully extended bazel to use `pip`. See: https://github.com/tanin47/bazel-test

One more thing left to test is reading the resource file.

**Comment by tanin47** [comment]
The full working example (with pip, pex, and using resource file) is here: https://github.com/tanin47/bazel-pex-pip

I'll integrate bazel into clusterfuzz-tools tomorrow.

**Comment by tanin47** [comment]
After 3 days of exploration, we will go with Pants instead of Bazel. See the explanation here: http://tanin.nanakorn.com/blogs/371

TL;DR Bazel doesn't support 3 necessary things (e.g. code coverage, pip, and pex). Pants supports all of them out-of-the-box. When Bazel supports these 3 things, it shouldn't take more than a day to switch over.

**Comment by tanin47** [comment]
https://github.com/google/clusterfuzz-tools/pull/166 is working on CI.

There are a few more follow-up items:
* Rename the folder `ci/`. Pants' test coverage gets confused and accidentally reports the module in `ci/continuous_integration/ci/tests`.
* We should make a target for Ansible execution. The system environment proves to be problematic.

**Comment by tanin47** [comment]
Ansible can't be integrated into Pants because it's not a pure python application. It uses a bunch of bash script. Therefore, it cannot be run hermetically.

Pants is working now; it can create the Pex for clusterfuzz which seems to work well.
